docs: rewrite for clarity, correct typos, grammer, etc (#4131)
yarnpkg/berry: packages/gatsby/content/getting-started/questions-and-answers.md

While the Yarn Classic line (1.x) remains a pillar of the JavaScript ecosystem, we recommend upgrading if possible. Why's that?

1. New features: On top of the features in Classic that you're already used to, you'll discover new features not found in Classic, including [`yarn dlx`](/cli/dlx), [builtin `patch:` protocol](https://github.com/yarnpkg/berry/tree/master/packages/plugin-patch), and more. Plus, new features can be, and are, added all the time through Modern's plugin system. Plugins extend Yarn's feature set with

2. Efficiency: Modern features new install strategies, leading to significantly smaller project sizes than is possible with Classic. For example, under Modern's default configuration, the stock CRA artifacts now only take 45MB instead of 237MB. [Performance](https://p.datadoghq.eu/sb/d2wdprp9uki7gfks-c562c42f4dfd0ade4885690fa719c818) has  improved as well, with most installs now only taking a few seconds to complete, even on extremely large projects. We even made it possible to reach [zero seconds](/features/zero-installs)!

> 3. Extensibility: Modern's architecture allows you to build your own features whenever you need. No more being blocked as you wait for a feature to be released by the core team. In fact, many of Modern's most popular features are implemented as plugins, including [changesets](/features/release-workflow), [constraints](/features/constraints), [workspaces](/cli/workspaces/foreach), and many more.  Focused workspaces, custom installs, project validation, all these and more are possible through the plugin system.

Reviewer: ```suggestion
3. Extensibility: Modern's architecture allows you to build your own features whenever you need to. No more being blocked as you wait for a feature to be released by the core team! In fact, many of Yarn's most popular features are implemented as plugins, including [changesets](/features/release-workflow), [constraints](/features/constraints), [workspaces](/cli/workspaces/foreach), and more. Focused workspaces, custom installs, project validation - all these and more are possible through the plugin system.
```

4. Stability: Modern builds on the years of experience spent maintaining Classic; it allowed us to finally fix longstanding design issues with how some features were implemented. Workspaces are now core components, the resolution pipeline has been streamlined, data structures are more efficient, and as a result, Modern is much less likely to suffer from incorrect assumptions and other design flaws.

> 5. Future-proof: A big reason why we invested in Modern was that we noticed how building new features on Classic was becoming difficult - each change leading to more and more unforeseen consequences. The Modern architecture corrects our past mistakes, and was designed to allow us to build features at a much higher pace.

Reviewer: ```suggestion
5. Future-proof: A big reason why we invested in Modern was that we noticed how building new features on Classic was becoming difficult - each change lead to more and more unforeseen consequences. Modern's architecture corrects our past mistakes and was designed to allow us to build features at a much higher pace.
```

## How easy should you expect the migration from Classic to Modern to be?

Generally, a few main things will need to be taken care of:

> 1. The settings format changed. We don't read the `.npmrc` or `.yarnrc` files anymore, instead we read settings from the [`.yarnrc.yml` file](https://yarnpkg.com/configuration/yarnrc).

Reviewer: ```suggestion
1. The settings format changed. We don't read the `.npmrc` or `.yarnrc` files anymore. Instead, we read settings from the [`.yarnrc.yml` file](https://yarnpkg.com/configuration/yarnrc).
```

- `.yarn/install-state.gz` is an optimization file that you shouldn't ever have to commit to your repository. It stores the exact state of your project so that the next commands can boot without having to resolve your workspaces all over again.

- `.yarn/patches` contains the patch files generated with the [`yarn patch-commit`](/cli/patch-commit) command. You should commit these files to your repository since they are necessary to install your dependencies.

- `.yarn/plugins` and `.yarn/releases` contain the Yarn releases used in the current repository (as defined by [`yarn set version`](/cli/set/version)). You should commit these files to your repository as this will prevent potential issues if, for example, two engineers use different Yarn versions with different features.

- `.yarn/sdks` contains the editor SDKs generated by `@yarnpkg/sdks`. You can decide whether to commit these to your repository or not. If you don't commit them, you'll need to follow the editor procedure again on new clones. See [Editor SDKs](/getting-started/editor-sdks) for more details.

> - `.yarn/unplugged` should usually be ignored since they typically hold machine-specific build artifacts. However, not committing them may prevent [Zero-Installs](https://yarnpkg.com/features/zero-installs) from working as intended. In that case, you can resolve this issue by setting [`enableScripts`](/configuration/yarnrc#enableScripts) to `false`.

Reviewer: ```suggestion
- The files in `.yarn/unplugged` should usually be ignored as they typically hold machine-specific build artifacts. However, not committing them may prevent [Zero-Installs](https://yarnpkg.com/features/zero-installs) from working as intended. In that case, you can resolve this issue by setting [`enableScripts`](/configuration/yarnrc#enableScripts) to `false`.
```

Unlike CommonJS, the ESM module resolution pipeline was intended to be completely walled from the outside so that, for example, multiple threads could share the same resolver instance. It also meant that without official loader support, **it was impossible to support alternate resolution strategies**. Monkey-patching the resolution primitives wasn't viable anymore, so none of those projects could support ESM at all. It could only mean one thing: **ESM wasn't ready**. Yes, it was supported natively, but given it broke a sizeable part of the ecosystem with no alternative whatsoever, it couldn't be a reasonable standard - yet.

> Fortunately, Node saw the issue, started to work on loaders, and shipped a first iteration. Fast forward to today, Node Loaders are still in [heavy work](https://nodejs.org/api/esm.html#loaders) (and changed shape more than once, as highlighted by this "experimental" annotation), but have allowed us to draft a first implementation of an ESM-compatible PnP loader, which we shipped in 3.1. Strong off those learnings, we started to contribute to the Node Loaders working group, not only to help Yarn's own use cases but also those from other projects susceptible to follow our lead.

Reviewer: ```suggestion
Fortunately, Node saw the issue, started to work on loaders, and shipped their first iteration. Fast forward to today, and while Node Loaders are still [a work in progress](https://nodejs.org/api/esm.html#loaders) (and have changed shape more than once, as highlighted by the"experimental" annotation), they have allowed us to draft the first implementation of an ESM-compatible PnP loader, which we shipped in 3.1. Strong off those learnings, we started to contribute to the Node Loaders working group, not only to help Yarn's use cases but also those of other projects that follow our lead.
```

Although compelling, this reasoning has a fatal flaw: removing the lockfile from the repository doesn't prevent this problem from happening. In particular:

- Active contributors won't get new versions unless they explicitly remove their install artifacts (`node_modules`), which may not happen often. Problematic dependency upgrades will thus be mainly discovered by new contributors, which doesn't make for a good first experience and may deter contributions.

- Even assuming you run fresh installs every week, your upgrades won't be easily reversible - once you test the most recent packages, you won't test against the less recent ones. The compatibility issues will still exist, they just will be against packages that used to work but that you don't test anymore. in other words, by always testing the most recent semver release, you won't see if you accidentally start relying on a feature that wasn't available before.

> Of course these points are only part of the problem - the lack of lockfile also means that key state information are missing from the repository. When months later you or your contributors want to make a fix on one of your old projects you might not even be able to *build* it anymore, let alone improve it.

Reviewer: ```suggestion
Of course, these points are only part of the problem - the lack of a lockfile also means that key state information is missing from the repository. When months later you or your contributors want to make a fix on one of your old projects you might not even be able to *build* it anymore, let alone improve it.
```

> Lockfiles should **always** be kept within the repository. Continuous integration testing **is a good idea**, but should be left to continuous integration systems. For example, Yarn itself runs [daily tests](https://github.com/yarnpkg/berry#current-status) against the latest versions of major open-source frameworks and tools, which allows us to quickly spot any compatibility issue with the newest release, while still being guaranteed that every contributor will have a consistent experience working with the project. [Dependabot](https://dependabot.com/#how-it-works) and [Renovate](https://www.whitesourcesoftware.com/free-developer-tools/renovate) are also good tools that track your dependencies updates for you.

Reviewer: ```suggestion
Lockfiles should **always** be kept within the repository. Continuous integration testing **is a good idea**, but should be left to continuous integration systems. For example, Yarn itself runs [daily tests](https://github.com/yarnpkg/berry#current-status) against the latest versions of major open-source frameworks and tools, which allows us to quickly spot any compatibility issues with the newest release, while still guaranteeing that every contributor will have a consistent experience working with the project. [Dependabot](https://dependabot.com/#how-it-works) and [Renovate](https://www.whitesourcesoftware.com/free-developer-tools/renovate) are also good tools that track dependency updates for you.
```

**No.**

Despite the first version of Yarn having been implemented by [Sebastian McKenzie](https://twitter.com/sebmck) while working at Facebook, the initial design received feedback from various other companies (such as [Tilde](https://www.tilde.io) via [Yehuda Katz](https://yehudakatz.com/2016/10/11/im-excited-to-work-on-yarn-the-new-js-package-manager-2/)) and the project was put into its own [GitHub organization](https://github.com/yarnpkg). Facebook kept investing in it during the following years (mostly because it proved to be a critical part of the React Native ecosystem) but major contributions came from the open-source community as well.

> Nowadays, the active development team is composed exclusively of people employed by non-founders companies. Facebook employees are of course still welcome to offer contributions to the project, but all contributions, no matter who they're from, go through the same review process.

Reviewer: ```suggestion
Nowadays, the active development team is composed exclusively of people employed by other companies. Facebook employees are of course still welcome to offer contributions to the project, but all contributions, no matter who they're from, go through the same review process.
```

## Why `registry.yarnpkg.com`? Does Facebook track us?

**No.**

At the time Yarn was created, the npm registry had issues that negatively impacted install performance, so the initial team decided to partner with Cloudflare and set up a [reverse proxy](https://en.wikipedia.org/wiki/Reverse_proxy) that would better cache the requests before returning them. This setup didn't even have a backend on our side.

> Eventually, npm resolved many of those issues, and we turned off the proxy to replace it by a [CNAME](https://en.wikipedia.org/wiki/CNAME_record) ([proof](https://toolbox.googleapps.com/apps/dig/#CNAME/registry.yarnpkg.com)). We still keep the hostname for because it gives us the ability to redirect to a read-only copy of the registry should the primary source become unavailable.

Reviewer: ```suggestion
Eventually, npm resolved many of those issues. We turned off the proxy and replaced it with a [CNAME](https://en.wikipedia.org/wiki/CNAME_record) ([proof](https://toolbox.googleapps.com/apps/dig/#CNAME/registry.yarnpkg.com)). We still keep the hostname as it gives us the ability to redirect to a read-only copy of the registry should the primary source become unavailable.
```

## Queries to `registry.yarnpkg.com` return a 404/500/...; is it down?

**No.**

As mentioned in the [previous section](#why-registryyarnpkgcom-does-facebook-track-us), the Yarn registry is just a CNAME to the npm registry. Since we don't even have a backend, any server error comes from the npm registry and thus should be reported to them and monitored on their [status page](https://status.npmjs.org/).

## Is Yarn faster than other package managers?

**Shrug 🤷‍♀️**

Yarn was much faster than some of its competitors when it was first released. Unfortunately, we failed to highlight that performance wasn't the main reason why we kept working on Yarn. Performance is an ongoing battle for most projects though, and Yarn's speed wasn't so much because we were doing something incredibly well, but rather the competing implementations had issues that seriously impacted their performance. As those issues were resolved, our miscommunication became more apparent as some people thought that Yarn was all about performances.

> Put simply, different projects make different tradeoffs.  We've spent significant resources pushing for Plug'n'Play (including through [dozens of contributions to third-party projects](https://github.com/pulls?utf8=%E2%9C%93&q=is%3Apr+author%3Aarcanis+archived%3Afalse+is%3Aclosed+pnp+-user%3Ayarnpkg+)) because we felt like this was important for the ecosystem. We prioritized workspaces because we felt like monorepos were providing significant value. This is the main difference: we make our own informed decisions regarding the project roadmap.

Reviewer: ```suggestion
Put simply, different projects make different tradeoffs. We've spent significant resources pushing for Plug'n'Play (including through [dozens of contributions to third-party projects](https://github.com/pulls?utf8=%E2%9C%93&q=is%3Apr+author%3Aarcanis+archived%3Afalse+is%3Aclosed+pnp+-user%3Ayarnpkg+)) because we felt like this was important for the ecosystem. We prioritized workspaces because we felt that monorepos provided significant value. This is Yarn's main distinction: we make our own informed decisions regarding the project roadmap.
```

Given that PnP is a resolver standard different from Node, tools that reimplement the `require.resolve` API need to add some logic to account for the PnP resolution. While various projects did so (for example Webpack 5 now supports PnP out of the box), a few are still on the fence about it. In the case of TypeScript we started and keep maintaining a [pull request](https://github.com/microsoft/TypeScript/pull/35206), but the TypeScript team still has to accept it. In order to unblock our users, we made the decision to automatically apply this exact pull request to the downloaded TypeScript versions, using our new [`patch:` protocol](/features/protocols#patch).

Which now begs the question: why do we still apply this patch even when Plug'n'Play is disabled? The main reason is that Yarn intends to provide consistent behaviour. Some setups involve using the `node_modules` linker during development (to avoid having to set up their editor [SDKS](/getting-started/editor-sdks)) and PnP in production (for install speed). If we were to only apply the patches when PnP is enabled, then the package cache would be different, which would break immutable installs.

> We *could have* potentially make it configurable through a switch, but in the end we decided it wasn't worth the extra configuration:

Reviewer: ```suggestion
We *could have* potentially made it configurable through a switch, but eventually we decided it wasn't worth the extra configuration:
```

> - The TypeScript patch is a noop if PnP isn't enabled, so this shouldn't affect your work (if it does, please open an issue)

Reviewer: ```suggestion
- The TypeScript patch is a no-op if PnP isn't enabled, so this shouldn't affect your workflow (if it does, please open an issue)
```

> - We hope to eventually land this PR in TypeScript one day, so the more eyes we can get on it the higher our confidence will be

Reviewer: ```suggestion
- We hope to eventually land this PR in TypeScript one day, so the more eyes we can get on it the more likely it will be that it will land
```

> - Since Yarn 3+, failing builtin patches are simply ignored and fallback to the original sources.

Reviewer: ```suggestion
- Beginning with Yarn 3, built-in patches that fail are simply ignored and Yarn falls back to the original package code
```
